clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java

```diff
@@ -843,7 +845,15 @@ private void updatePatternSubscription(Cluster cluster) {
     @Override
     public void unsubscribe() {
         fetchBuffer.retainAll(Collections.emptySet());
-        subscriptions.unsubscribe();
+        UnsubscribeApplicationEvent unsubscribeApplicationEvent = new UnsubscribeApplicationEvent();
+        applicationEventHandler.add(unsubscribeApplicationEvent);
+        unsubscribeApplicationEvent.future().whenComplete((result, error) -> {
+            if (error != null) {
+                // Callback failed - Keeping same exception message thrown by the legacy consumer
+                throw new KafkaException("User rebalance callback throws an error", error);
```

Comment:
Looking at the current consumer, I think it can be quite complicated in this implementation.  Maybe what we should do is to invoke the listener on the spot, then send an even to the background thread to leave group.

Reply:
You're right, it should happen in the poll to maintain the current contract. I will just remove it since we don't execute callbacks yet, and it should be included in the PR that introduces the callback execution.

Reply:
The current consumer throws KafkaException on unsubscribe(), I think throwing in the whenComplete doesn't actually throw.

Reply:
I think unsubscribe() actually blocks on callback invocation and throw if possible.  Instead of putting the logic in whenComplete, it seems like we should try to wait till the callback completes then throw if needed.  I assume we want to maintain this behavior for the async consumer.

Reply:
I think this KafkaException might not be in the right place because the rebalance listener needs to be invoked on the mainthread, probably before sending out the event.  I wonder if we could just remove this whenComplete and rely on the background thread to log the failures during the leave group event.  If there's a fatal exception being thrown there, it seems the sensible way is to enqueue to the BackgroundEventQueue and handle in the poll.  wdyt?

In the current code path, I think only exceptions can only be thrown in `onLeavePrepare`.

Reply:
Agree that the unsubscribe blocks on the callbacks, but since we don't have the implementation for how callbacks are going to be executed on this PR this unsubscribe is still not using it. It should come when we nail the implementation details on the follow-up PR (will depend on how we end up doing it, maybe blocking not here, but on the subscribe/unsubscribe events)

As for the exception, it will originate in the Application thread, where the callback is executed, and it should only be returned to the user when it calls poll, to maintain the current behaviour, so I removed it from here to stay consistent and leave all logic related to the callback execution out, I see the confusion introduced.

The when complete should stay because it represents a concept we need, un-related to callbacks: we do need to update the `subscriptionState` only when the Unsubscribe event completes (HB to leave group sent to the broker). We need it know to make sure we are able to run unsubscribe, no callbacks, sending leave group, and clearing up the subscription state after sending the request. 

Trying to leave it in a consistent state with no callbacks. Follow-up PR should introduce implementation for executing them, blocking appropriately on the execution, and throwing the exceptions.